### PR TITLE
Add Thai segmenter

### PR DESCRIPTION
**Pull request by aFluffyHotdog**
# Pull Request
- Implement `ThaiSegmenter`
- add tests
- add benchmarks

## What does this PR do?
Fixes [#113](https://github.com/meilisearch/charabia/issues/113)

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

**Comment by ManyTheFish** [comment]
Hello @aFluffyHotdog! 
Your work seems promising, do you have any issue finishing it?

**Comment by aFluffyHotdog** [comment]
@ManyTheFish Oh I just forgot to package the dictionary required for the tokenizer to work, I'll reopen the pull request.

**Comment by ManyTheFish** [comment]
No problem! Ping me when the PR is ready or if you have any question 😄

File: benches/bench.rs [hunk 1]
```diff
@@ -11,14 +11,15 @@ static DATA_SET: &[((usize, Script, Language), &str)] = &[
...
     ((363, Script::Latin, Language::Fra), "La position de Lutèce, sur l'île aujourd'hui nommée l'île de la Cité, permettant le franchissement du grand fleuve navigable qu'est la Seine par une voie reliant le Nord et le Sud des Gaules, en fait dès l'Antiquité une cité importante, capitale des Parisii, puis lieu de séjour d'un empereur romain. Le mot Lutèce resulte de la francisation de Lutetia."),
     ((365, Script::Hebrew, Language::Heb), "הַשּׁוּעָל הַמָּהִיר (״הַחוּם״) לֹא יָכוֹל לִקְפֹּץ 8.94 מֶטְרִים, נָכוֹן? תַּכְלֶס, אִם הוּא הָיָה יָכוֹל, הוּא חֲתִיכַת שׁוּעָל הַשּׁוּעָל הַזֶּה.. אֲבָל הַאִם לֹא כֻּלָּנוּ שׁוּעָלִים בְּעֶצֶם? יתכן."),
-];
+    ((132, Script::Thai, Language::Tha), "เราจะทำตามสัญญาขอเวลาอีกไม่นานแล้วแผ่นดินที่งดงามจะคืนกลับมาเราจะทำอย่างซื่อตรงขอแค่เธอจงไว้ใจและศรัทธาแผ่นดินจะดีในไม่ช้า"),
```
Comment (ManyTheFish) [inline benches/bench.rs:21]:
```suggestion
    ((366, Script::Thai, Language::Tha), "เราจะทำตามสัญญาขอเวลาอีกไม่นานแล้วแผ่นดินที่งดงามจะคืนกลับมาเราจะทำอย่างซื่อตรงขอแค่เธอจงไว้ใจและศรัทธาแผ่นดินจะดีในไม่ช้า"),
```

**Comment by aFluffyHotdog** [comment]
> No problem! Ping me when the PR is ready or if you have any questions 😄
> 
> I would like to know the source of your dictionary, can you add an URL somewhere in the implementation?
> 
> In another hand, I would like to warn you about the content of the dictionary, it contains really long terms like `กรมสมเด็จพระเทพรัตนราชสุดาฯ สยามบรมราชกุมารี` and the library that you use will segment the text matching the longest term found. Is it expected for `กรมสมเด็จพระเทพรัตนราชสุดาฯ สยามบรมราชกุมารี` to be considered as a unique word?

So the dictionary came from PyThaiNLP (a thai NLP library for python). Regarding that word, it is considered to be unique, it is the name of the queen which might explain why it was added into the dictionary.

**Comment by aFluffyHotdog** [comment]
Ready for review! :D @ManyTheFish

**Comment by ManyTheFish** [comment]
Hello @aFluffyHotdog, thank you for your PR!

Could you make some changes before we merge it to the main branch:
- Make a cargo fmt pass, in order to format the code
- Move the dictionary in a directory at the root of the repository
- Fix the tests running `cargo test`

Thank you again!

**Comment by aFluffyHotdog** [comment]
Got it!

**Comment by aFluffyHotdog** [comment]
@ManyTheFish  All good to go!

image

File: src/segmenter/thai.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
+// Import `Segmenter` trait.
+use crate::segmenter::Segmenter;
+use once_cell::sync::Lazy;
+use std::{ffi::OsStr, fs, path::Path};
```
Comment (ManyTheFish) [inline src/segmenter/thai.rs:4]:
```suggestion
```

**Review by ManyTheFish — CHANGES_REQUESTED** [review]

**Comment by ManyTheFish** [comment]
Thank you @aFluffyHotdog!
Bors merge

**Review by ManyTheFish — APPROVED** [review]

**Comment by ManyTheFish** [comment]
bors merge

**Pull request merged**